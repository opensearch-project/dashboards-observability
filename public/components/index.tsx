/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { QueryManager } from 'common/query_manager';
import React from 'react';
import ReactDOM from 'react-dom';
import { AppMountParameters, CoreStart } from '../../../../src/core/public';
import { DataSourceManagementPluginSetup } from '../../../../src/plugins/data_source_management/public';
import { AppPluginStartDependencies } from '../types';
import { App } from './app';

export const Observability = (
  CoreStartProp: CoreStart,
  DepsStart: AppPluginStartDependencies,
  AppMountParametersProp: AppMountParameters,
  pplService: any,
  dslService: any,
  savedObjects: any,
  timestampUtils: any,
  queryManager: QueryManager,
  startPage: string,
  dataSourcePluggables,
  dataSourceManagement: DataSourceManagementPluginSetup,
  savedObjectsMDSClient: CoreStart['savedObjects']
) => {
  const { setHeaderActionMenu } = AppMountParametersProp;
  const { dataSource } = DepsStart;
  ReactDOM.render(
    <App
      CoreStartProp={CoreStartProp}
      DepsStart={DepsStart}
      pplService={pplService}
      dslService={dslService}
      savedObjects={savedObjects}
      timestampUtils={timestampUtils}
      queryManager={queryManager}
      startPage={startPage}
      dataSourcePluggables={dataSourcePluggables}
      dataSourceManagement={dataSourceManagement}
      setActionMenu={setHeaderActionMenu}
      dataSourceEnabled={!!dataSource}
      savedObjectsMDSClient={savedObjectsMDSClient}
    />,
    AppMountParametersProp.element
  );

  return () => ReactDOM.unmountComponentAtNode(AppMountParametersProp.element);
};
