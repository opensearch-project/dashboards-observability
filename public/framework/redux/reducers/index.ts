/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { combineReducers } from 'redux';

import { queriesReducer } from '../../../components/event_analytics/redux/slices/query_slice';
import { queryResultsReducer } from '../../../components/event_analytics/redux/slices/query_result_slice';
import { queryTabReducer } from '../../../components/event_analytics/redux/slices/query_tab_slice';
import { fieldsReducer } from '../../../components/event_analytics/redux/slices/field_slice';
import { countDistributionReducer } from '../../../components/event_analytics/redux/slices/count_distribution_slice';
import { explorerVisualizationReducer } from '../../../components/event_analytics/redux/slices/visualization_slice';
import { explorerVisualizationConfigReducer } from '../../../components/event_analytics/redux/slices/viualization_config_slice';
import { patternsReducer } from '../../../components/event_analytics/redux/slices/patterns_slice';
import { metricsReducers } from '../../../components/metrics/redux/slices/metrics_slice';
import { panelReducer } from '../../../components/custom_panels/redux/panel_slice';
import { searchMetaDataSliceReducer } from '../../../components/event_analytics/redux/slices/search_meta_data_slice';

const combinedReducer = combineReducers({
  // explorer reducers
  queries: queriesReducer,
  queryResults: queryResultsReducer,
  explorerTabs: queryTabReducer,
  fields: fieldsReducer,
  countDistribution: countDistributionReducer,
  explorerVisualization: explorerVisualizationReducer,
  explorerVisualizationConfig: explorerVisualizationConfigReducer,
  patterns: patternsReducer,
  metrics: metricsReducers,
  customPanel: panelReducer,
  searchMetadata: searchMetaDataSliceReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export const rootReducer = combinedReducer;
