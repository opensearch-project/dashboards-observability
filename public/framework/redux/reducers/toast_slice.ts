import { createSlice } from '@reduxjs/toolkit';

const toastSlice = createSlice({
  name: 'toast',
  initialState: { toasts: [], toastRightSide: true },
  reducers: {
    appendToast: (state, action) => {
      state.toasts = [...state.toasts, action.payload];
    },

    setToastRightSide: (state, action) => (state.toastRightSide = action.payload),

    dismissToast: (state, action) => {
      state.toasts = state.toasts.filter((t) => t.id !== action.payload.id);
    },

    resetToasts: (state, action) => {
      state.toasts = [];
      state.toastRightSide = true;
    },
  },
});

export const toastReducer = toastSlice.reducer;

export const { dismissToast, resetToasts } = toastSlice.actions;
const { appendToast, setToastRightSide } = toastSlice.actions;

export const selectToasts = (rootState) => rootState.toast.toasts;

export const selectToastRightSide = (rootState) => rootState.toast.toastRightSide;

export const addToast = (title, color?, textChild?, side?) => (dispatch, getState) => {
  const newToast = { id: new Date().toISOString(), title, textChild, color };
  dispatch(appendToast(newToast));

  if (side) {
    dispatch(setToastRightSide(side === 'left'));
  }
};
