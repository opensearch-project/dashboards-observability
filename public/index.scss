/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

/* base */

@import 'variables';

/* components */

// event analytics
@import 'components/event_analytics/index';