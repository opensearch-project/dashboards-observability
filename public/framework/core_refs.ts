/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import DSLService from 'public/services/requests/dsl';
import {
  ApplicationStart,
  ChromeStart,
  CoreStart,
  HttpStart,
  IToasts,
  OverlayStart,
  SavedObjectsClientContract,
} from '../../../../src/core/public';
import { DashboardStart } from '../../../../src/plugins/dashboard/public';
import { DataSourcePluginStart } from '../../../../src/plugins/data_source/public';
import PPLService from '../services/requests/ppl';

class CoreRefs {
  private static _instance: CoreRefs;

  public core?: CoreStart;
  public http?: HttpStart;
  public savedObjectsClient?: SavedObjectsClientContract;
  public pplService?: PPLService;
  public dslService?: DSLService;
  public toasts?: IToasts;
  public chrome?: ChromeStart;
  public application?: ApplicationStart;
  public queryAssistEnabled?: boolean;
  public summarizeEnabled?: boolean;
  public dashboard?: DashboardStart;
  public dashboardProviders?: unknown;
  public overlays?: OverlayStart;
  public dataSource?: DataSourcePluginStart;
  private constructor() {
    // ...
  }

  public static get Instance() {
    // Do you need arguments? Make it a regular static method instead.
    return this._instance || (this._instance = new this());
  }
}

export const coreRefs = CoreRefs.Instance;
