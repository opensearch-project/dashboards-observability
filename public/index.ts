/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { PluginInitializerContext } from '../../../src/core/public';
import './components/notebooks/index.scss';
import './components/trace_analytics/index.scss';
import { ObservabilityPlugin } from './plugin';
import './variables.scss';

export { ObservabilityPlugin as Plugin };

export const plugin = (initializerContext: PluginInitializerContext) =>
  new ObservabilityPlugin(initializerContext);

export { ObservabilityStart } from './types';
