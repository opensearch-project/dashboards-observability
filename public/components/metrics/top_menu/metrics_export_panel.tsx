/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiComboBox,
  EuiFieldText,
  EuiFlexGroup,
  EuiFlexItem,
  EuiForm,
  EuiFormRow,
  EuiHorizontalRule,
  EuiSpacer,
  EuiText,
  EuiTitle,
} from '@elastic/eui';
import React, { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { MetricType } from '../../../../common/types/metrics';
import { fetchPanels } from '../../../../public/components/custom_panels/redux/panel_slice';

interface MetricsExportPanelProps {
  metricsToExport: MetricType[];
  setMetricsToExport: React.Dispatch<React.SetStateAction<MetricType[]>>;
  availableDashboards: any[];
  selectedPanelOptions: any[];
  setSelectedPanelOptions: React.Dispatch<React.SetStateAction<any[]>>;
}

interface CustomPanelOptions {
  id: string;
  name: string;
  dateCreated: string;
  dateModified: string;
}

export const MetricsExportPanel = ({
  metricsToExport,
  setMetricsToExport,
  availableDashboards,
  selectedPanelOptions,
  setSelectedPanelOptions,
}: MetricsExportPanelProps) => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchPanels());
  }, [dispatch]);

  const onNameChange = (index: number, name: string) => {
    const tempMetrics = metricsToExport.map((metric, idx) =>
      idx === index ? { ...metric, name } : metric
    );
    setMetricsToExport(tempMetrics);
  };

  return (
    <div style={{ minWidth: '15vw' }}>
      <EuiTitle size="xs">
        <h3>SAVE THE VIEW AS VISUALIZATION</h3>
      </EuiTitle>
      <EuiHorizontalRule margin="s" />

      {metricsToExport.length > 0 && (
        <div
          style={{
            maxHeight: '30vh',
            overflowY: 'scroll',
            width: 'auto',
            overflowX: 'hidden',
            marginBottom: '16px',
          }}
        >
          {metricsToExport.map((metaData: any, index: number) => {
            return (
              <EuiForm component="form" key={`save-panel-id-${index}`}>
                <EuiFlexGroup>
                  <EuiFlexItem>
                    <EuiFormRow label={'Metric ' + (index + 1)}>
                      <EuiFieldText
                        key={`metric-name-input-id-${index}`}
                        value={metaData.name}
                        onChange={(e) => onNameChange(index, e.target.value)}
                        data-test-subj="metrics__querySaveName"
                      />
                    </EuiFormRow>
                  </EuiFlexItem>
                </EuiFlexGroup>
              </EuiForm>
            );
          })}
        </div>
      )}

      <EuiFlexGroup direction="column" gutterSize="none">
        {!(metricsToExport.length > 0) && (
          <EuiFlexItem style={{ maxWidth: '400px' }}>
            <EuiText size="s" color="subdued">
              <p>
                Save the view as visualization. You can add it to custom operational dashboards or
                applications.
              </p>
            </EuiText>
            <EuiSpacer size="m" />
          </EuiFlexItem>
        )}

        {metricsToExport.length > 0 && (
          <EuiFlexItem>
            <EuiText size="s">
              <p>
                <strong>Add to custom operational dashboards or applications.</strong>
              </p>
            </EuiText>
          </EuiFlexItem>
        )}

        <EuiFlexItem style={{ maxWidth: '400px' }}>
          <EuiFormRow label="Dashboards and applications - optional">
            <EuiComboBox
              placeholder="Select dashboards/applications"
              onChange={(newOptions) => {
                setSelectedPanelOptions(newOptions);
              }}
              selectedOptions={selectedPanelOptions}
              options={
                availableDashboards?.map((option: any) => ({
                  panel: option,
                  label: option.title,
                })) ?? []
              }
              isClearable={true}
              data-test-subj="eventExplorer__querySaveComboBox"
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
    </div>
  );
};
