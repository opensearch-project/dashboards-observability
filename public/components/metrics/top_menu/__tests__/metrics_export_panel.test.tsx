/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import httpClientMock from '../../../../../test/__mocks__/httpClientMock';
import {
  samplePanelOptions,
  sampleSortedMetricsLayout,
  sampleVisualizationById,
} from '../../../../../test/metrics_contants';
import { applyMiddleware, createStore } from '@reduxjs/toolkit';
import { rootReducer } from '../../../../framework/redux/reducers';
import { Provider } from 'react-redux';
import { HttpResponse } from '../../../../../../../src/core/public';
import { MetricsExportPanel } from '../metrics_export_panel';
import { EuiComboBoxOptionOption } from '@elastic/eui';
import thunk from 'redux-thunk';
import { coreRefs } from '../../../../framework/core_refs';

describe('Export Metrics Panel Component', () => {
  configure({ adapter: new Adapter() });
  const store = createStore(rootReducer, applyMiddleware(thunk));
  coreRefs.savedObjectsClient.find = jest.fn(() =>
    Promise.resolve({
      savedObjects: [],
      then: () => Promise.resolve(),
    })
  );

  it('renders Export Metrics Panel Component', async () => {
    let httpFlag = 1;
    httpClientMock.get = jest.fn(() => {
      if (httpFlag === 1) {
        httpFlag += 1;
        return Promise.resolve((samplePanelOptions as unknown) as HttpResponse);
      } else {
        return Promise.resolve((sampleVisualizationById as unknown) as HttpResponse);
      }
    });

    const http = httpClientMock;
    const visualizationsMetaData: any = [];
    const setVisualizationsMetaData = jest.fn();
    const sortedMetricsLayout = sampleSortedMetricsLayout;
    const selectedPanelOptions: Array<EuiComboBoxOptionOption<unknown>> = [];
    const setSelectedPanelOptions = jest.fn();

    const wrapper = mount(
      <Provider store={store}>
        <MetricsExportPanel
          http={http}
          visualizationsMetaData={visualizationsMetaData}
          setVisualizationsMetaData={setVisualizationsMetaData}
          sortedMetricsLayout={sortedMetricsLayout}
          selectedPanelOptions={selectedPanelOptions}
          setSelectedPanelOptions={setSelectedPanelOptions}
        />
      </Provider>
    );
    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
