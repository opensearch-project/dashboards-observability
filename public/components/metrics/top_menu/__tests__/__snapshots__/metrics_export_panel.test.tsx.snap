// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export Metrics Panel Component renders Export Metrics Panel Component 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(Symbol.observable): [Function],
    }
  }
>
  <MetricsExportPanel
    availableDashboards={
      Array [
        Object {
          "attributes": Object {
            "title": "[Flights] Global Flight Dashboard",
          },
          "id": "7adfa750-4c81-11e8-b3d7-01146121b73d",
          "title": "[Flights] Global Flight Dashboard",
          "type": "dashboard",
        },
        Object {
          "id": "fdf2bb60-7a5b-11ee-929a-5f1a2dc08039",
          "name": "[Logs] Web traffic Panel",
          "objectId": "observability-panel:fdf2bb60-7a5b-11ee-929a-5f1a2dc08039",
          "savedObject": true,
          "title": "[Logs] Web traffic Panel",
          "type": "observability-panel",
        },
      ]
    }
    metricsToExport={
      Array [
        Object {
          "aggregation": "avg",
          "attributesGroupBy": Array [],
          "availableAttributes": Array [],
          "catalog": "my_prometheus",
          "catalogSourceName": "my_prometheus",
          "catalogTableName": "go_memstats_alloc_bytes",
          "id": "my_prometheus.go_memstats_alloc_bytes",
          "index": "my_prometheus.go_memstats_alloc_bytes",
          "name": "my_prometheus.go_memstats_alloc_bytes",
          "recentlyCreated": false,
          "sub_type": "promqlmetric",
          "type": "line",
        },
        Object {
          "catalog": "CUSTOM_METRICS",
          "id": "observability-visualization:84c73aa0-84aa-11ee-96e0-7bfa0b41d0fc",
          "name": "my_prometheus.go_memstats_alloc_bytes_total",
          "query": "source = my_prometheus.query_range('sum by(instance,job) (go_memstats_alloc_bytes_total)', 1700071440, 1700157840, '1h')",
          "recentlyCreated": false,
          "savedVisualizationId": "observability-visualization:84c73aa0-84aa-11ee-96e0-7bfa0b41d0fc",
          "type": "line",
        },
      ]
    }
  >
    <div
      style={
        Object {
          "minWidth": "15vw",
        }
      }
    >
      <EuiFormRow
        describedByIds={Array []}
        display="row"
        fullWidth={false}
        hasChildLabel={true}
        hasEmptyLabelSpace={false}
        helpText="Search existing dashboards or applications by name"
        label="Custom operational dashboards/application"
        labelType="label"
      >
        <div
          className="euiFormRow"
          id="random_html_id-row"
        >
          <div
            className="euiFormRow__labelWrapper"
          >
            <EuiFormLabel
              className="euiFormRow__label"
              htmlFor="random_html_id"
              isFocused={false}
              type="label"
            >
              <label
                className="euiFormLabel euiFormRow__label"
                htmlFor="random_html_id"
              >
                Custom operational dashboards/application
              </label>
            </EuiFormLabel>
          </div>
          <div
            className="euiFormRow__fieldWrapper"
          >
            <EuiComboBox
              aria-describedby="random_html_id-help-0"
              async={false}
              compressed={false}
              data-test-subj="eventExplorer__querySaveComboBox"
              fullWidth={false}
              id="random_html_id"
              isClearable={true}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              options={
                Array [
                  Object {
                    "label": "[Flights] Global Flight Dashboard",
                    "panel": Object {
                      "attributes": Object {
                        "title": "[Flights] Global Flight Dashboard",
                      },
                      "id": "7adfa750-4c81-11e8-b3d7-01146121b73d",
                      "title": "[Flights] Global Flight Dashboard",
                      "type": "dashboard",
                    },
                  },
                  Object {
                    "label": "[Logs] Web traffic Panel",
                    "panel": Object {
                      "id": "fdf2bb60-7a5b-11ee-929a-5f1a2dc08039",
                      "name": "[Logs] Web traffic Panel",
                      "objectId": "observability-panel:fdf2bb60-7a5b-11ee-929a-5f1a2dc08039",
                      "savedObject": true,
                      "title": "[Logs] Web traffic Panel",
                      "type": "observability-panel",
                    },
                  },
                ]
              }
              placeholder="Select dashboards/applications"
              selectedOptions={Array []}
              singleSelection={false}
              sortMatchesBy="none"
            >
              <div
                aria-describedby="random_html_id-help-0"
                aria-expanded={false}
                aria-haspopup="listbox"
                className="euiComboBox"
                data-test-subj="eventExplorer__querySaveComboBox"
                onKeyDown={[Function]}
                role="combobox"
              >
                <EuiComboBoxInput
                  autoSizeInputRef={[Function]}
                  compressed={false}
                  fullWidth={false}
                  hasSelectedOptions={false}
                  id="random_html_id"
                  inputRef={[Function]}
                  isListOpen={false}
                  noIcon={false}
                  onChange={[Function]}
                  onClear={[Function]}
                  onClick={[Function]}
                  onCloseListClick={[Function]}
                  onFocus={[Function]}
                  onOpenListClick={[Function]}
                  onRemoveOption={[Function]}
                  placeholder="Select dashboards/applications"
                  rootId={[Function]}
                  searchValue=""
                  selectedOptions={Array []}
                  singleSelection={false}
                  toggleButtonRef={[Function]}
                  updatePosition={[Function]}
                  value=""
                >
                  <EuiFormControlLayout
                    compressed={false}
                    fullWidth={false}
                    icon={
                      Object {
                        "aria-label": "Open list of options",
                        "data-test-subj": "comboBoxToggleListButton",
                        "disabled": undefined,
                        "onClick": [Function],
                        "ref": [Function],
                        "side": "right",
                        "type": "arrowDown",
                      }
                    }
                  >
                    <div
                      className="euiFormControlLayout"
                    >
                      <div
                        className="euiFormControlLayout__childrenWrapper"
                      >
                        <div
                          className="euiComboBox__inputWrap euiComboBox__inputWrap-isClearable"
                          data-test-subj="comboBoxInput"
                          onClick={[Function]}
                          tabIndex={-1}
                        >
                          <p
                            className="euiComboBoxPlaceholder"
                          >
                            Select dashboards/applications
                          </p>
                          <AutosizeInput
                            aria-controls=""
                            className="euiComboBox__input"
                            data-test-subj="comboBoxSearchInput"
                            id="random_html_id"
                            injectStyles={true}
                            inputRef={[Function]}
                            minWidth={1}
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            role="textbox"
                            style={
                              Object {
                                "fontSize": 14,
                              }
                            }
                            value=""
                          >
                            <div
                              className="euiComboBox__input"
                              style={
                                Object {
                                  "display": "inline-block",
                                  "fontSize": 14,
                                }
                              }
                            >
                              <input
                                aria-controls=""
                                data-test-subj="comboBoxSearchInput"
                                id="random_html_id"
                                onBlur={[Function]}
                                onChange={[Function]}
                                onFocus={[Function]}
                                role="textbox"
                                style={
                                  Object {
                                    "boxSizing": "content-box",
                                    "width": "2px",
                                  }
                                }
                                value=""
                              />
                              <div
                                style={
                                  Object {
                                    "height": 0,
                                    "left": 0,
                                    "overflow": "scroll",
                                    "position": "absolute",
                                    "top": 0,
                                    "visibility": "hidden",
                                    "whiteSpace": "pre",
                                  }
                                }
                              />
                            </div>
                          </AutosizeInput>
                        </div>
                        <EuiFormControlLayoutIcons
                          compressed={false}
                          icon={
                            Object {
                              "aria-label": "Open list of options",
                              "data-test-subj": "comboBoxToggleListButton",
                              "disabled": undefined,
                              "onClick": [Function],
                              "ref": [Function],
                              "side": "right",
                              "type": "arrowDown",
                            }
                          }
                        >
                          <div
                            className="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                          >
                            <EuiFormControlLayoutCustomIcon
                              aria-label="Open list of options"
                              data-test-subj="comboBoxToggleListButton"
                              iconRef={[Function]}
                              onClick={[Function]}
                              size="m"
                              type="arrowDown"
                            >
                              <button
                                aria-label="Open list of options"
                                className="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                                data-test-subj="comboBoxToggleListButton"
                                onClick={[Function]}
                                type="button"
                              >
                                <EuiIcon
                                  aria-hidden="true"
                                  className="euiFormControlLayoutCustomIcon__icon"
                                  size="m"
                                  type="arrowDown"
                                >
                                  <EuiIconBeaker
                                    aria-hidden={true}
                                    className="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                                    focusable="false"
                                    role="img"
                                    style={null}
                                  >
                                    <svg
                                      aria-hidden={true}
                                      className="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                                      focusable="false"
                                      height={16}
                                      role="img"
                                      style={null}
                                      viewBox="0 0 16 16"
                                      width={16}
                                      xmlns="http://www.w3.org/2000/svg"
                                    >
                                      <path
                                        d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                                      />
                                    </svg>
                                  </EuiIconBeaker>
                                </EuiIcon>
                              </button>
                            </EuiFormControlLayoutCustomIcon>
                          </div>
                        </EuiFormControlLayoutIcons>
                      </div>
                    </div>
                  </EuiFormControlLayout>
                </EuiComboBoxInput>
              </div>
            </EuiComboBox>
            <EuiFormHelpText
              className="euiFormRow__text"
              id="random_html_id-help-0"
              key="Search existing dashboards or applications by name"
            >
              <div
                className="euiFormHelpText euiFormRow__text"
                id="random_html_id-help-0"
              >
                Search existing dashboards or applications by name
              </div>
            </EuiFormHelpText>
          </div>
        </div>
      </EuiFormRow>
      <div
        style={
          Object {
            "maxHeight": "30vh",
            "overflowX": "hidden",
            "overflowY": "scroll",
            "width": "auto",
          }
        }
      >
        <EuiForm
          component="form"
          key="save-panel-id-0"
        >
          <form
            className="euiForm"
          >
            <EuiFlexGroup>
              <div
                className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
              >
                <EuiFlexItem>
                  <div
                    className="euiFlexItem"
                  >
                    <EuiFormRow
                      describedByIds={Array []}
                      display="row"
                      fullWidth={false}
                      hasChildLabel={true}
                      hasEmptyLabelSpace={false}
                      label="Metric Name #1"
                      labelType="label"
                    >
                      <div
                        className="euiFormRow"
                        id="random_html_id-row"
                      >
                        <div
                          className="euiFormRow__labelWrapper"
                        >
                          <EuiFormLabel
                            className="euiFormRow__label"
                            htmlFor="random_html_id"
                            isFocused={false}
                            type="label"
                          >
                            <label
                              className="euiFormLabel euiFormRow__label"
                              htmlFor="random_html_id"
                            >
                              Metric Name #1
                            </label>
                          </EuiFormLabel>
                        </div>
                        <div
                          className="euiFormRow__fieldWrapper"
                        >
                          <EuiFieldText
                            data-test-subj="metrics__querySaveName"
                            id="random_html_id"
                            key="metric-name-input-id-0"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            value="my_prometheus.go_memstats_alloc_bytes"
                          >
                            <EuiFormControlLayout
                              fullWidth={false}
                              inputId="random_html_id"
                            >
                              <div
                                className="euiFormControlLayout"
                              >
                                <div
                                  className="euiFormControlLayout__childrenWrapper"
                                >
                                  <EuiValidatableControl>
                                    <input
                                      className="euiFieldText"
                                      data-test-subj="metrics__querySaveName"
                                      id="random_html_id"
                                      onBlur={[Function]}
                                      onChange={[Function]}
                                      onFocus={[Function]}
                                      type="text"
                                      value="my_prometheus.go_memstats_alloc_bytes"
                                    />
                                  </EuiValidatableControl>
                                  <EuiFormControlLayoutIcons />
                                </div>
                              </div>
                            </EuiFormControlLayout>
                          </EuiFieldText>
                        </div>
                      </div>
                    </EuiFormRow>
                  </div>
                </EuiFlexItem>
              </div>
            </EuiFlexGroup>
          </form>
        </EuiForm>
        <EuiForm
          component="form"
          key="save-panel-id-1"
        >
          <form
            className="euiForm"
          >
            <EuiFlexGroup>
              <div
                className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
              >
                <EuiFlexItem>
                  <div
                    className="euiFlexItem"
                  >
                    <EuiFormRow
                      describedByIds={Array []}
                      display="row"
                      fullWidth={false}
                      hasChildLabel={true}
                      hasEmptyLabelSpace={false}
                      label="Metric Name #2"
                      labelType="label"
                    >
                      <div
                        className="euiFormRow"
                        id="random_html_id-row"
                      >
                        <div
                          className="euiFormRow__labelWrapper"
                        >
                          <EuiFormLabel
                            className="euiFormRow__label"
                            htmlFor="random_html_id"
                            isFocused={false}
                            type="label"
                          >
                            <label
                              className="euiFormLabel euiFormRow__label"
                              htmlFor="random_html_id"
                            >
                              Metric Name #2
                            </label>
                          </EuiFormLabel>
                        </div>
                        <div
                          className="euiFormRow__fieldWrapper"
                        >
                          <EuiFieldText
                            data-test-subj="metrics__querySaveName"
                            id="random_html_id"
                            key="metric-name-input-id-1"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            value="my_prometheus.go_memstats_alloc_bytes_total"
                          >
                            <EuiFormControlLayout
                              fullWidth={false}
                              inputId="random_html_id"
                            >
                              <div
                                className="euiFormControlLayout"
                              >
                                <div
                                  className="euiFormControlLayout__childrenWrapper"
                                >
                                  <EuiValidatableControl>
                                    <input
                                      className="euiFieldText"
                                      data-test-subj="metrics__querySaveName"
                                      id="random_html_id"
                                      onBlur={[Function]}
                                      onChange={[Function]}
                                      onFocus={[Function]}
                                      type="text"
                                      value="my_prometheus.go_memstats_alloc_bytes_total"
                                    />
                                  </EuiValidatableControl>
                                  <EuiFormControlLayoutIcons />
                                </div>
                              </div>
                            </EuiFormControlLayout>
                          </EuiFieldText>
                        </div>
                      </div>
                    </EuiFormRow>
                  </div>
                </EuiFlexItem>
              </div>
            </EuiFlexGroup>
          </form>
        </EuiForm>
      </div>
    </div>
  </MetricsExportPanel>
</Provider>
`;
